File: tests/udi_tests/singlestep.cxx
/*
 * Copyright (c) 2011-2015, UDI Contributors
 * All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <cstdlib>
#include <cstring>
#include <iostream>
#include <sstream>

#include "libudi.h"
#include "libuditest.h"
#include "test_bins.h"
#include "test_lib.h"

using std::cout;
using std::endl;
using std::stringstream;
using std::string;

class test_singlestep : public UDITestCase {
    public:
        test_singlestep()
            : UDITestCase(std::string("test_singlestep")) {}
        virtual ~test_singlestep() {}

        bool operator()(void);
};

static const char *TEST_BINARY = SIMPLE_BINARY_PATH;
static udi_address TEST_FUNCTION = SIMPLE_FUNCTION2;
static unsigned long TEST_FUNCTION_LEN = SIMPLE_FUNCTION2_LENGTH;

static test_singlestep testInstance;

bool test_singlestep::operator()(void) {
    char *argv[] = { NULL };

    udi_proc_config config;
    config.root_dir = NULL;

    udi_error_e error_code;
    char *errmsg = NULL;
    udi_process *proc = create_process(TEST_BINARY, argv, NULL, &config, &error_code, &errmsg);
    free(errmsg);

    test_assert(proc != NULL);

    udi_thread *thr = get_initial_thread(proc);
    test_assert(thr != NULL);

    udi_error_e result = create_breakpoint(proc, TEST_FUNCTION);
    assert_no_error(proc, result);

    result = install_breakpoint(proc, TEST_FUNCTION);
    assert_no_error(proc, result);

    result = continue_process(proc);
    assert_no_error(proc, result);

    wait_for_breakpoint(thr, TEST_FUNCTION);

    // Enable single stepping
    result = set_single_step(thr, 1);
    assert_no_error(proc, result);
    test_assert(get_single_step(thr) == 1);

    udi_address current_pc;
    result = get_pc(thr, &current_pc);
    assert_no_error(proc, result);
    test_assert(current_pc == TEST_FUNCTION);

    // Single step through the whole test function
    udi_address next_pc;
    do {
        result = continue_process(proc);
        assert_no_error(proc, result);

        wait_for_single_step(thr);

        result = get_next_instruction(thr, &next_pc);
        assert_no_error(proc, result);

        result = get_pc(thr, &current_pc);
        assert_no_error(proc, result);
    }while (    (current_pc < (TEST_FUNCTION + TEST_FUNCTION_LEN))
             && (next_pc < (TEST_FUNCTION + TEST_FUNCTION_LEN)) );

    // Disable single stepping
    result = set_single_step(thr, 0);
    assert_no_error(proc, result);
    test_assert(get_single_step(thr) == 0);

    result = continue_process(proc);
    assert_no_error(proc, result);

    wait_for_exit(thr, EXIT_FAILURE);

    return true;
}

File: libudi/tests/singlestep.rs
//
// Copyright (c) 2011-2017, UDI Contributors
// All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
#![deny(warnings)]

extern crate udi;

#[macro_use]
extern crate lazy_static;

mod native_file_tests;
mod utils;

use udi::Result;

#[test]
fn singlestep() {
    if let Err(e) = singlestep_test() {
        utils::print_error(e);
        panic!("singlestep test failed");
    }
}

fn singlestep_test() -> Result<()> {
    let addr = native_file_tests::get_test_metadata().simple_function2_addr();
    let len = native_file_tests::get_test_metadata().simple_function2_length();
    let exec_path = native_file_tests::get_test_metadata().simple_path().to_str().unwrap();

    let config = udi::ProcessConfig::new(None, utils::rt_lib_path());
    let argv = Vec::new();
    let envp = Vec::new();

    let proc_ref = udi::create_process(exec_path,
                                       &argv,
                                       &envp,
                                       &config)?;
    let thr_ref;
    {
        let mut process = proc_ref.lock()?;

        thr_ref = process.get_initial_thread();
        process.create_breakpoint(addr)?;
        process.install_breakpoint(addr)?;
        process.continue_process()?;
    }

    utils::wait_for_event(&proc_ref, &thr_ref, &udi::EventData::Breakpoint{ addr });

    let mut current_pc = thr_ref.lock()?.get_pc()?;
    assert_eq!(addr, current_pc);

    thr_ref.lock()?.set_single_step(true)?;

    // single step through the whole test function
    let mut next_pc: u64 = thr_ref.lock()?.get_next_instruction()?;
    while next_pc < addr + len && next_pc >= addr &&
          current_pc < addr + len && current_pc >= addr {

        proc_ref.lock()?.continue_process()?;

        utils::wait_for_event(&proc_ref, &thr_ref, &udi::EventData::SingleStep);

        next_pc = thr_ref.lock()?.get_next_instruction()?;

        current_pc = thr_ref.lock()?.get_pc()?;
    }

    // disable single stepping
    thr_ref.lock()?.set_single_step(false)?;

    proc_ref.lock()?.continue_process()?;

    utils::wait_for_exit(&proc_ref, &thr_ref, 1);

    Ok(())
}
